### PR TITLE
filter out completion results that don't start with the replacement text

### PR DESCRIPTION
**Pull request by johnpryan**
I'm not sure if this is going to work for all scenarios... I've done some basic testing with the new embeddable editor

Screen Shot 2019-04-30 at 3 24 47 PM

closes dart-lang/dart-pad/#1014

### CODE REVIEW

File: test/analysis_server_test.dart [hunk 1]
```diff
@@ -176,6 +183,20 @@ void defineTests() {
       AnalysisResults results = await analysisServer.analyze(sampleDart2OK);
       expect(results.issues, hasLength(0));
     });
+
+    test('filter completions', () async {
+      // just after A
+      var idx = 61;
+      expect(completionLargeNamespaces.substring(idx - 1, idx), 'A');
+      return analysisServer
```
Comment (devoncarew) [inline test/analysis_server_test.dart]: use `await` here instead of `then`?

File: lib/src/analysis_server.dart [hunk 2]
```diff
@@ -332,7 +332,10 @@ class AnalysisServerWrapper {
       );
       CompletionResults results = await getCompletionResults(id.id);
       await _unloadSources();
-      return results;
+      var source = sources[_getPathFromName(sourceName)];
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: Perhaps move the filtering to `completeMulti()`? You then wouldn't have to create a new copy of `CompletionResults`, which is really an impl class from the analysis server library.

File: lib/src/analysis_server.dart [hunk 3]
```diff
@@ -332,7 +332,10 @@ class AnalysisServerWrapper {
...
+      var prefix = source.substring(results.replacementOffset,
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: For `prefix`, we may want the string from `results.replacementOffset` to `offset`. `replacementLength` can extend past the cursor position I believe.

https://htmlpreview.github.io/?https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/doc/api.html, under 'completion.results'.

File: lib/src/analysis_server.dart [hunk 4]
```diff
@@ -125,10 +125,13 @@ class AnalysisServerWrapper {
         await _completeImpl(sources, location.sourceName, location.offset);
     List<CompletionSuggestion> suggestions = results.results;
 
-    // This hack filters out of scope completions. It needs removing when we
-    // have categories of completions.
-    // TODO(devoncarew): Remove this filter code.
+    var source = sources[location.sourceName];
+    var prefix = source.substring(results.replacementOffset, location.offset);
     suggestions = suggestions
+        .where((s) => s.completion.startsWith(prefix))
```
Comment (devoncarew) [inline lib/src/analysis_server.dart]: we may want to compare case insensitive here (via toLowerCase on each completion, and once on `prefix`)

**Comment by johnpryan** [comment]
added the case-insensitive piece. it's certainly fast on when the server is running locally :)

**Pull request merged**